ui/StatusQ/src/wallet/managetokenscontroller.h

```diff
@@ -0,0 +1,92 @@
+#include <QObject>
+#include <QSettings>
+#include <QQmlParserStatus>
+
+#include <array>
+
+#include "managetokensmodel.h"
+
+class QAbstractItemModel;
+
+class ManageTokensController : public QObject, public QQmlParserStatus
+{
+    Q_OBJECT
+    Q_INTERFACES(QQmlParserStatus)
+
+    // input properties
+    Q_PROPERTY(QAbstractItemModel* sourceModel READ sourceModel WRITE setSourceModel NOTIFY sourceModelChanged FINAL REQUIRED)
+    Q_PROPERTY(QString settingsKey READ settingsKey WRITE setSettingsKey NOTIFY settingsKeyChanged FINAL REQUIRED)
+    Q_PROPERTY(bool arrangeByCommunity READ arrangeByCommunity WRITE setArrangeByCommunity NOTIFY arrangeByCommunityChanged FINAL)
+
+    // output properties
+    Q_PROPERTY(QAbstractItemModel* regularTokensModel READ regularTokensModel CONSTANT FINAL)
+    // TODO regularTokenGroupsModel for grouped (collections of) collectibles?
+    Q_PROPERTY(QAbstractItemModel* communityTokensModel READ communityTokensModel CONSTANT FINAL)
+    Q_PROPERTY(QAbstractItemModel* communityTokenGroupsModel READ communityTokenGroupsModel CONSTANT FINAL)
+    Q_PROPERTY(QAbstractItemModel* hiddenTokensModel READ hiddenTokensModel CONSTANT FINAL)
+    Q_PROPERTY(bool dirty READ dirty NOTIFY dirtyChanged FINAL)
+
+public:
+    explicit ManageTokensController(QObject* parent = nullptr);
+
+    Q_INVOKABLE void showHideRegularToken(int row, bool flag);
+    Q_INVOKABLE void showHideCommunityToken(int row, bool flag);
+    Q_INVOKABLE void showHideGroup(const QString& groupId, bool flag);
+
+    Q_INVOKABLE void saveSettings();
+    Q_INVOKABLE void clearSettings();
+    Q_INVOKABLE void revert();
+
+    // TODO: to be used by SFPM on the main wallet page
+    // bool lessThan(lhsSymbol, rhsSymbol) const;
+    // bool filterAcceptsRow(index or symbol?) const;
+
+protected:
+    void classBegin() override;
+    void componentComplete() override;
+
+signals:
+    void sourceModelChanged();
+    void dirtyChanged();
+    void arrangeByCommunityChanged();
+    void settingsKeyChanged();
+
+private:
+    QAbstractItemModel* m_sourceModel{nullptr};
```

Comment:
Not sure about this one...

Reply:
The thing is that `m_sourceModel` will be received from the outside and ManageTokensController has no control over the model lifetime. ManageTokensController could easily remain with a zombie pointer.
So I think there's two actions to take when this happens:
1. Clean the `m_sourceModel`. This can be simplified by using `QPointer`
2. Notify whoever has a binding on `ManageTokensController::sourceModel` property that the model has changed.

Reply:
OK, fixed (eventhough w/o a QPointer which I find cumbersome to use in this scenario)